src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs

```diff
@@ -334,35 +343,44 @@ private void UpdateStartingStreams(DateTimeOffset now)
             }
             finally
             {
+                // TODO should this message say the connection is shutting down or closing rather than faulted?
+                // Faulted has a negative meaning and we can get here by the host stopping or ConnectionLifeTimeNotification.RequestClose.
                 var connectionError = error as ConnectionAbortedException
                     ?? new ConnectionAbortedException(CoreStrings.Http3ConnectionFaulted, error!);
 
                 try
                 {
-                    if (TryClose())
+                    // Only send goaway if the connection close was initiated on the server.
+                    if (!clientAbort)
                     {
-                        // This throws when connection is shut down.
-                        // TODO how to make it so we can distinguish between Abort from server vs client?
-                        await SendGoAway(GetHighestStreamId());
+                        await SendGoAwayAsync(GetCurrentGoAwayStreamId());
```

Comment:
Couldn't this GoAway already have been sent by UpdateConnectionState?

Reply:
Yeah. It's actually desirable if we sent long.MaxValue in the previous GOAWAY and this one then sends the exact value. There might be a situation where the exact value is sent twice. That doesn't break the spec, but it's not something we _want_ to do.

I'd like to fix that up later. I'll see if I can create that situation in a test and fix then.

Reply:
With the current logic it looks like you'd always get two final GoAway's if the shutdown happened without any active streams.